In [ ]:
import pandas as pd

In [ ]:
links= pd.read_csv('/content/links.csv')
movies= pd.read_csv('/content/movies.csv')
ratings= pd.read_csv('/content/ratings.csv')
tags= pd.read_csv('/content/tags.csv')



Links



In [ ]:
links.shape

(9742, 3)

In [ ]:
links.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   movieId  9742 non-null   int64  
 1   imdbId   9742 non-null   int64  
 2   tmdbId   9734 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 228.5 KB


In [ ]:
links.describe()

movieId        imdbId         tmdbId
count    9742.000000  9.742000e+03    9734.000000
mean    42200.353623  6.771839e+05   55162.123793
std     52160.494854  1.107228e+06   93653.481487
min         1.000000  4.170000e+02       2.000000
25%      3248.250000  9.518075e+04    9665.500000
50%      7300.000000  1.672605e+05   16529.000000
75%     76232.000000  8.055685e+05   44205.750000
max    193609.000000  8.391976e+06  525662.000000

In [ ]:
links.nunique()

movieId    9742
imdbId     9742
tmdbId     9733
dtype: int64

In [ ]:
links.isna()

movieId  imdbId  tmdbId
0       False   False   False
1       False   False   False
2       False   False   False
3       False   False   False
4       False   False   False
...       ...     ...     ...
9737    False   False   False
9738    False   False   False
9739    False   False   False
9740    False   False   False
9741    False   False   False

[9742 rows x 3 columns]

In [ ]:
links.isna().sum()

movieId    0
imdbId     0
tmdbId     8
dtype: int64

In [ ]:
links.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

Movies

In [ ]:
movies.shape

(9742, 3)

In [ ]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  9742 non-null   int64 
 1   title    9742 non-null   object
 2   genres   9742 non-null   object
dtypes: int64(1), object(2)
memory usage: 228.5+ KB


In [ ]:
movies.describe()

movieId
count    9742.000000
mean    42200.353623
std     52160.494854
min         1.000000
25%      3248.250000
50%      7300.000000
75%     76232.000000
max    193609.000000

In [ ]:
movies.nunique()

movieId    9742
title      9737
genres      951
dtype: int64

In [ ]:
movies.isna()

movieId  title  genres
0       False  False   False
1       False  False   False
2       False  False   False
3       False  False   False
4       False  False   False
...       ...    ...     ...
9737    False  False   False
9738    False  False   False
9739    False  False   False
9740    False  False   False
9741    False  False   False

[9742 rows x 3 columns]

In [ ]:
movies.isna().sum()

movieId    0
title      0
genres     0
dtype: int64

In [ ]:
movies.title.duplicated().sum()

5

In [ ]:
movies.head(25)

movieId                                  title  \
0         1                       Toy Story (1995)   
1         2                         Jumanji (1995)   
2         3                Grumpier Old Men (1995)   
3         4               Waiting to Exhale (1995)   
4         5     Father of the Bride Part II (1995)   
5         6                            Heat (1995)   
6         7                         Sabrina (1995)   
7         8                    Tom and Huck (1995)   
8         9                    Sudden Death (1995)   
9        10                       GoldenEye (1995)   
10       11         American President, The (1995)   
11       12     Dracula: Dead and Loving It (1995)   
12       13                           Balto (1995)   
13       14                           Nixon (1995)   
14       15                Cutthroat Island (1995)   
15       16                          Casino (1995)   
16       17           Sense and Sensibility (1995)   
17       18                      Four Rooms (1995)   
18       19  Ace Ventura: When Nature Calls (1995)   
19       20                     Money Train (1995)   
20       21                      Get Shorty (1995)   
21       22                         Copycat (1995)   
22       23                       Assassins (1995)   
23       24                          Powder (1995)   
24       25               Leaving Las Vegas (1995)   

                                         genres  
0   Adventure|Animation|Children|Comedy|Fantasy  
1                    Adventure|Children|Fantasy  
2                                Comedy|Romance  
3                          Comedy|Drama|Romance  
4                                        Comedy  
5                         Action|Crime|Thriller  
6                                Comedy|Romance  
7                            Adventure|Children  
8                                        Action  
9                     Action|Adventure|Thriller  
10                         Comedy|Drama|Romance  
11                                Comedy|Horror  
12                 Adventure|Animation|Children  
13                                        Drama  
14                     Action|Adventure|Romance  
15                                  Crime|Drama  
16                                Drama|Romance  
17                                       Comedy  
18                                       Comedy  
19           Action|Comedy|Crime|Drama|Thriller  
20                        Comedy|Crime|Thriller  
21          Crime|Drama|Horror|Mystery|Thriller  
22                        Action|Crime|Thriller  
23                                 Drama|Sci-Fi  
24                                Drama|Romance

Ratings

In [ ]:
ratings.shape

(100836, 4)

In [ ]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   userId     100836 non-null  int64  
 1   movieId    100836 non-null  int64  
 2   rating     100836 non-null  float64
 3   timestamp  100836 non-null  int64  
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


In [ ]:
ratings.describe()

userId        movieId         rating     timestamp
count  100836.000000  100836.000000  100836.000000  1.008360e+05
mean      326.127564   19435.295718       3.501557  1.205946e+09
std       182.618491   35530.987199       1.042529  2.162610e+08
min         1.000000       1.000000       0.500000  8.281246e+08
25%       177.000000    1199.000000       3.000000  1.019124e+09
50%       325.000000    2991.000000       3.500000  1.186087e+09
75%       477.000000    8122.000000       4.000000  1.435994e+09
max       610.000000  193609.000000       5.000000  1.537799e+09

In [ ]:
ratings.nunique()

userId         610
movieId       9724
rating          10
timestamp    85043
dtype: int64

In [ ]:
ratings.isna().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

In [ ]:
ratings.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

Tags

In [ ]:
tags.shape

(3683, 4)

In [ ]:
tags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3683 entries, 0 to 3682
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   userId     3683 non-null   int64 
 1   movieId    3683 non-null   int64 
 2   tag        3683 non-null   object
 3   timestamp  3683 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 115.2+ KB


In [ ]:
tags.describe()

userId        movieId     timestamp
count  3683.000000    3683.000000  3.683000e+03
mean    431.149335   27252.013576  1.320032e+09
std     158.472553   43490.558803  1.721025e+08
min       2.000000       1.000000  1.137179e+09
25%     424.000000    1262.500000  1.137521e+09
50%     474.000000    4454.000000  1.269833e+09
75%     477.000000   39263.000000  1.498457e+09
max     610.000000  193565.000000  1.537099e+09

In [ ]:
tags.nunique()

userId         58
movieId      1572
tag          1589
timestamp    3411
dtype: int64

In [ ]:
tags.isna().sum()

userId       0
movieId      0
tag          0
timestamp    0
dtype: int64

In [ ]:
tags.head()

userId  movieId              tag   timestamp
0       2    60756            funny  1445714994
1       2    60756  Highly quotable  1445714996
2       2    60756     will ferrell  1445714992
3       2    89774     Boxing story  1445715207
4       2    89774              MMA  1445715200

In [ ]:
# Find Lowest and Highest rated movies:
mean_rating = ratings.groupby('movieId')[['rating']].mean()
mean_rating

rating
movieId          
1        3.920930
2        3.431818
3        3.259615
4        2.357143
5        3.071429
...           ...
193581   4.000000
193583   3.500000
193585   3.500000
193587   3.500000
193609   4.000000

[9724 rows x 1 columns]

In [ ]:
#Lowest rated movies
lowest_rated = mean_rating['rating'].idxmin()
movies.loc[movies['movieId'] == lowest_rated]

movieId         title   genres
2689     3604  Gypsy (1962)  Musical

In [ ]:
highest_rated = mean_rating['rating'].idxmax()
movies.loc[movies['movieId'] == highest_rated]

movieId            title           genres
48       53  Lamerica (1994)  Adventure|Drama

In [ ]:
# show number of people who rated movies highest
ratings[ratings['movieId']==highest_rated]

userId  movieId  rating  timestamp
13368      85       53     5.0  889468268
96115     603       53     5.0  963180003

In [ ]:
# show number of people who rated movies lowest
ratings[ratings['movieId']==lowest_rated]

userId  movieId  rating   timestamp
13633      89     3604     0.5  1520408880

In [ ]:
# merging both the datasets on 'movieId' column
movie_rating=movies.merge(ratings,on='movieId',how='left')
movie_rating

movieId                                      title  \
0             1                           Toy Story (1995)   
1             1                           Toy Story (1995)   
2             1                           Toy Story (1995)   
3             1                           Toy Story (1995)   
4             1                           Toy Story (1995)   
...         ...                                        ...   
100849   193581  Black Butler: Book of the Atlantic (2017)   
100850   193583               No Game No Life: Zero (2017)   
100851   193585                               Flint (2017)   
100852   193587        Bungo Stray Dogs: Dead Apple (2018)   
100853   193609        Andrew Dice Clay: Dice Rules (1991)   

                                             genres  userId  rating  \
0       Adventure|Animation|Children|Comedy|Fantasy     1.0     4.0   
1       Adventure|Animation|Children|Comedy|Fantasy     5.0     4.0   
2       Adventure|Animation|Children|Comedy|Fantasy     7.0     4.5   
3       Adventure|Animation|Children|Comedy|Fantasy    15.0     2.5   
4       Adventure|Animation|Children|Comedy|Fantasy    17.0     4.5   
...                                             ...     ...     ...   
100849              Action|Animation|Comedy|Fantasy   184.0     4.0   
100850                     Animation|Comedy|Fantasy   184.0     3.5   
100851                                        Drama   184.0     3.5   
100852                             Action|Animation   184.0     3.5   
100853                                       Comedy   331.0     4.0   

           timestamp  
0       9.649827e+08  
1       8.474350e+08  
2       1.106636e+09  
3       1.510578e+09  
4       1.305696e+09  
...              ...  
100849  1.537109e+09  
100850  1.537110e+09  
100851  1.537110e+09  
100852  1.537110e+09  
100853  1.537158e+09  

[100854 rows x 6 columns]

In [ ]:
movie_rating.sort_values('rating', ascending=False).head(5)


movieId                title                        genres  userId  \
56255     3676    Eraserhead (1977)                  Drama|Horror   567.0   
33889     1639   Chasing Amy (1997)          Comedy|Drama|Romance   147.0   
20742      914  My Fair Lady (1964)  Comedy|Drama|Musical|Romance   188.0   
10168      356  Forrest Gump (1994)      Comedy|Drama|Romance|War   283.0   
10169      356  Forrest Gump (1994)      Comedy|Drama|Romance|War   284.0   

       rating     timestamp  
56255     5.0  1.525282e+09  
33889     5.0  1.203268e+09  
20742     5.0  9.625602e+08  
10168     5.0  9.012280e+08  
10169     5.0  8.326953e+08

In [ ]:
avg_rating_movies = movie_rating.groupby('title')['rating'].mean().sort_values(ascending=False)
avg_rating_movies.head(5)

title
Gena the Crocodile (1969)              5.0
True Stories (1986)                    5.0
Cosmic Scrat-tastrophe (2015)          5.0
Love and Pigeons (1985)                5.0
Red Sorghum (Hong gao liang) (1987)    5.0
Name: rating, dtype: float64

In [ ]:
# grouping the movie_rating based on count on userId and mean on rating(sorting by rating)
rating_mean= movie_rating.groupby('title')[['userId','rating']].agg({'userId':'count','rating':'mean'}).round(2).sort_values(by='rating',ascending=False)
rating_mean.head()

userId  rating
title                                                             
A Perfect Day (2015)                                     1     5.0
Last Year's Snow Was Falling (1983)                      1     5.0
Sun Alley (Sonnenallee) (1999)                           1     5.0
Four Days in September (O Que É Isso, Companhei...       1     5.0
The Love Bug (1997)                                      1     5.0

Popularity/Quality based recommmender system

In [ ]:
rating_popularity= pd.DataFrame(ratings.groupby('movieId')['rating'].mean())
rating_popularity.sort_values("rating", ascending=False).head()

rating
movieId        
88448       5.0
100556      5.0
143031      5.0
143511      5.0
143559      5.0

In [ ]:
rating_popularity.query("movieId==88448")

rating
movieId        
88448       5.0

In [ ]:
rating_popularity['rating_count'] = ratings.groupby('movieId')['rating'].count()
rating_popularity.sort_values("rating_count", ascending=False).head()

rating  rating_count
movieId                        
356      4.164134           329
318      4.429022           317
296      4.197068           307
593      4.161290           279
2571     4.192446           278

In [ ]:
Final_res=rating_popularity.merge(movies, how="inner", on="movieId")
Final_res

movieId    rating  rating_count  \
0           1  3.920930           215   
1           2  3.431818           110   
2           3  3.259615            52   
3           4  2.357143             7   
4           5  3.071429            49   
...       ...       ...           ...   
9719   193581  4.000000             1   
9720   193583  3.500000             1   
9721   193585  3.500000             1   
9722   193587  3.500000             1   
9723   193609  4.000000             1   

                                          title  \
0                              Toy Story (1995)   
1                                Jumanji (1995)   
2                       Grumpier Old Men (1995)   
3                      Waiting to Exhale (1995)   
4            Father of the Bride Part II (1995)   
...                                         ...   
9719  Black Butler: Book of the Atlantic (2017)   
9720               No Game No Life: Zero (2017)   
9721                               Flint (2017)   
9722        Bungo Stray Dogs: Dead Apple (2018)   
9723        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9719              Action|Animation|Comedy|Fantasy  
9720                     Animation|Comedy|Fantasy  
9721                                        Drama  
9722                             Action|Animation  
9723                                       Comedy  

[9724 rows x 5 columns]

In [ ]:
top_10=Final_res.sort_values("rating_count", ascending=False).head(10)

In [ ]:
top_10[['movieId','title','genres','rating','rating_count']]

movieId                                      title  \
314       356                        Forrest Gump (1994)   
277       318           Shawshank Redemption, The (1994)   
257       296                        Pulp Fiction (1994)   
510       593           Silence of the Lambs, The (1991)   
1938     2571                         Matrix, The (1999)   
224       260  Star Wars: Episode IV - A New Hope (1977)   
418       480                       Jurassic Park (1993)   
97        110                          Braveheart (1995)   
507       589          Terminator 2: Judgment Day (1991)   
461       527                    Schindler's List (1993)   

                                genres    rating  rating_count  
314           Comedy|Drama|Romance|War  4.164134           329  
277                        Crime|Drama  4.429022           317  
257        Comedy|Crime|Drama|Thriller  4.197068           307  
510              Crime|Horror|Thriller  4.161290           279  
1938            Action|Sci-Fi|Thriller  4.192446           278  
224            Action|Adventure|Sci-Fi  4.231076           251  
418   Action|Adventure|Sci-Fi|Thriller  3.750000           238  
97                    Action|Drama|War  4.031646           237  
507                      Action|Sci-Fi  3.970982           224  
461                          Drama|War  4.225000           220

In [ ]:
import pickle
pickle.dump(top_10, open(file='Recommended_movies.sav', mode='wb'))

Top movie is Forest Gump with rating count 329

Item based collaborative technique

In [ ]:
# creating pivot table to create item by item collaborative filtering
movie_rating_pivot = pd.pivot_table(data=ratings, values='rating', index='userId', columns='movieId')
#movie_rating_pivot.fillna(0, inplace=True)

In [ ]:
movie_rating_pivot.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     NaN     4.0     NaN     NaN     4.0     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  193583  193585  193587  193609  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN  

[5 rows x 9724 columns]

In [ ]:
top_popular_movieId =356

In [ ]:
FG_rating= movie_rating_pivot[top_popular_movieId] #fg is forest gump
FG_rating[FG_rating>=0]

userId
1      4.0
6      5.0
7      5.0
8      3.0
10     3.5
      ... 
605    3.0
606    4.0
608    3.0
609    4.0
610    3.0
Name: 356, Length: 329, dtype: float64

In [ ]:
similar_to_FG = movie_rating_pivot.corrwith(FG_rating)
similar_to_FG 

/usr/local/lib/python3.7/dist-packages/numpy/lib/function_base.py:2683: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar, dtype=dtype)
/usr/local/lib/python3.7/dist-packages/numpy/lib/function_base.py:2542: RuntimeWarning: divide by zero encountered in true_divide
  c *= np.true_divide(1, fact)


movieId
1         0.303465
2         0.367247
3         0.534682
4         0.388514
5         0.349541
            ...   
193581         NaN
193583         NaN
193585         NaN
193587         NaN
193609         NaN
Length: 9724, dtype: float64

In [ ]:
FG_correlation= pd.DataFrame(similar_to_FG, columns=['PearsonR'])
FG_correlation.dropna(inplace=True)
FG_correlation.head(10)

PearsonR
movieId          
1        0.303465
2        0.367247
3        0.534682
4        0.388514
5        0.349541
6        0.137421
7        0.106567
8        0.656020
9        0.000000
10       0.217441

In [ ]:
rating = pd.DataFrame(ratings.groupby('movieId')['rating'].mean())
rating['rating_count'] =ratings.groupby('movieId')['rating'].count()

In [ ]:
FG_corr_summary = FG_correlation.join(rating['rating_count'])
FG_corr_summary.drop(top_popular_movieId, inplace=True) 
FG_corr_summary

PearsonR  rating_count
movieId                        
1        0.303465           215
2        0.367247           110
3        0.534682            52
4        0.388514             7
5        0.349541            49
...           ...           ...
185585  -1.000000             2
187541   1.000000             4
187593  -0.203519            12
187595   0.870388             5
188301   0.693375             3

[5459 rows x 2 columns]

Picking the top 10 movies

In [ ]:
top10 = FG_corr_summary[FG_corr_summary['rating_count']>=10].sort_values('PearsonR', ascending=False).head(10)
top10

PearsonR  rating_count
movieId                        
1295     0.932958            11
6793     0.885253            11
328      0.881682            10
4954     0.865633            11
911      0.850591            13
55721    0.799415            10
195      0.786428            10
181      0.785661            17
80906    0.782601            12
150548   0.776636            10

In [ ]:
def top_n_movies(movieId, n):
    movie_ratings = movie_rating_pivot[movieId]
    similar_to_movies = movie_rating_pivot.corrwith(movie_ratings)
    corr_movies = pd.DataFrame(similar_to_movies, columns=['PearsonR'])
    corr_movies.dropna(inplace=True)
    movies_corr_summary = corr_movies.join(rating['rating_count'])
    movies_corr_summary.drop(movieId, inplace=True) # drop the inputed restaurant itself
    top10 = movies_corr_summary[movies_corr_summary['rating_count']>=10].sort_values('PearsonR', ascending=False).head(n)
    top10 = top10.merge(movies, left_index=True, right_on="movieId")
    return list(top10["title"])

In [ ]:
top_n_movies(356,10)

/usr/local/lib/python3.7/dist-packages/numpy/lib/function_base.py:2683: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar, dtype=dtype)
/usr/local/lib/python3.7/dist-packages/numpy/lib/function_base.py:2542: RuntimeWarning: divide by zero encountered in true_divide
  c *= np.true_divide(1, fact)


['Unbearable Lightness of Being, The (1988)',
 'Beethoven (1992)',
 'Tales from the Crypt Presents: Demon Knight (1995)',
 "Ocean's Eleven (a.k.a. Ocean's 11) (1960)",
 'Charade (1963)',
 'Elite Squad (Tropa de Elite) (2007)',
 'Something to Talk About (1995)',
 'Mighty Morphin Power Rangers: The Movie (1995)',
 'Inside Job (2010)',
 'Sherlock: The Abominable Bride (2016)']

User_based Collaborative technique

In [ ]:
# creating pivot table to create item by item collaborative filtering
user_pivot = pd.pivot_table(data=ratings, values='rating', index='userId', columns='movieId')
user_pivot.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     NaN     4.0     NaN     NaN     4.0     NaN     NaN   
2           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
5           4.0     NaN     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
2           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
3           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
4           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   
5           NaN     NaN  ...     NaN     NaN     NaN     NaN     NaN     NaN   

movieId  193583  193585  193587  193609  
userId                                   
1           NaN     NaN     NaN     NaN  
2           NaN     NaN     NaN     NaN  
3           NaN     NaN     NaN     NaN  
4           NaN     NaN     NaN     NaN  
5           NaN     NaN     NaN     NaN  

[5 rows x 9724 columns]

In [ ]:
#replacing NaNs with zeros
user_pivot.fillna(0, inplace=True)
user_pivot.head()

movieId  1       2       3       4       5       6       7       8       \
userId                                                                    
1           4.0     0.0     4.0     0.0     0.0     4.0     0.0     0.0   
2           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
5           4.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  9       10      ...  193565  193567  193571  193573  193579  193581  \
userId                   ...                                                   
1           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
2           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
3           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
4           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
5           0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

movieId  193583  193585  193587  193609  
userId                                   
1           0.0     0.0     0.0     0.0  
2           0.0     0.0     0.0     0.0  
3           0.0     0.0     0.0     0.0  
4           0.0     0.0     0.0     0.0  
5           0.0     0.0     0.0     0.0  

[5 rows x 9724 columns]

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

user_similarities = pd.DataFrame(cosine_similarity(user_pivot),
                                 columns=user_pivot.index, 
                                 index=user_pivot.index)
user_similarities.head()

userId       1         2         3         4         5         6         7    \
userId                                                                         
1       1.000000  0.027283  0.059720  0.194395  0.129080  0.128152  0.158744   
2       0.027283  1.000000  0.000000  0.003726  0.016614  0.025333  0.027585   
3       0.059720  0.000000  1.000000  0.002251  0.005020  0.003936  0.000000   
4       0.194395  0.003726  0.002251  1.000000  0.128659  0.088491  0.115120   
5       0.129080  0.016614  0.005020  0.128659  1.000000  0.300349  0.108342   

userId       8         9         10   ...       601       602       603  \
userId                                ...                                 
1       0.136968  0.064263  0.016875  ...  0.080554  0.164455  0.221486   
2       0.027257  0.000000  0.067445  ...  0.202671  0.016866  0.011997   
3       0.004941  0.000000  0.000000  ...  0.005048  0.004892  0.024992   
4       0.062969  0.011361  0.031163  ...  0.085938  0.128273  0.307973   
5       0.429075  0.000000  0.030611  ...  0.068048  0.418747  0.110148   

userId       604       605       606       607       608       609       610  
userId                                                                        
1       0.070669  0.153625  0.164191  0.269389  0.291097  0.093572  0.145321  
2       0.000000  0.000000  0.028429  0.012948  0.046211  0.027565  0.102427  
3       0.000000  0.010694  0.012993  0.019247  0.021128  0.000000  0.032119  
4       0.052985  0.084584  0.200395  0.131746  0.149858  0.032198  0.107683  
5       0.258773  0.148758  0.106435  0.152866  0.135535  0.261232  0.060792  

[5 rows x 610 columns]

In [ ]:
# compute the weights for one user
user_id = 1

weights = (
    user_similarities.query("userId!=@user_id")[user_id] / sum(user_similarities.query("userId!=@user_id")[user_id])
          )
weights.head(6)

userId
2    0.000336
3    0.000736
4    0.002395
5    0.001590
6    0.001579
7    0.001956
Name: 1, dtype: float64

In [ ]:
weights.sum()

1.000000000000001

In [ ]:
user_pivot.loc[user_id,:]==0

movieId
1         False
2          True
3         False
4          True
5          True
          ...  
193581     True
193583     True
193585     True
193587     True
193609     True
Name: 1, Length: 9724, dtype: bool

In [ ]:
# select movies that the inputed user has not been rated
not_rated_movies = user_pivot.loc[user_pivot.index!=user_id, user_pivot.loc[user_id,:]==0]
not_rated_movies.T

userId   2    3    4    5    6    7    8    9    10   11   ...  601  602  603  \
movieId                                                    ...                  
2        0.0  0.0  0.0  0.0  4.0  0.0  4.0  0.0  0.0  0.0  ...  0.0  4.0  0.0   
4        0.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5        0.0  0.0  0.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
7        0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
8        0.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
193581   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193583   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193585   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193587   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
193609   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

userId   604  605  606  607  608  609  610  
movieId                                     
2        5.0  3.5  0.0  0.0  2.0  0.0  0.0  
4        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
5        3.0  0.0  0.0  0.0  0.0  0.0  0.0  
7        0.0  0.0  2.5  0.0  0.0  0.0  0.0  
8        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...      ...  ...  ...  ...  ...  ...  ...  
193581   0.0  0.0  0.0  0.0  0.0  0.0  0.0  
193583   0.0  0.0  0.0  0.0  0.0  0.0  0.0  
193585   0.0  0.0  0.0  0.0  0.0  0.0  0.0  
193587   0.0  0.0  0.0  0.0  0.0  0.0  0.0  
193609   0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[9492 rows x 609 columns]

In [ ]:
# dot product between the not-visited-restaurants and the weights
weighted_averages = pd.DataFrame(not_rated_movies.T.dot(weights), columns=["predicted_rating"])
weighted_averages.head()

predicted_rating
movieId                  
2                0.842127
4                0.027652
5                0.275115
7                0.321403
8                0.046876

In [ ]:
recommendations = weighted_averages.merge(movies, left_index=True, right_on="movieId")
recommendations.sort_values("predicted_rating", ascending=False).head()

predicted_rating  movieId  \
277           2.654727      318   
507           2.087327      589   
659           1.859548      858   
2078          1.663564     2762   
3638          1.624820     4993   

                                                  title                genres  
277                    Shawshank Redemption, The (1994)           Crime|Drama  
507                   Terminator 2: Judgment Day (1991)         Action|Sci-Fi  
659                               Godfather, The (1972)           Crime|Drama  
2078                            Sixth Sense, The (1999)  Drama|Horror|Mystery  
3638  Lord of the Rings: The Fellowship of the Ring,...     Adventure|Fantasy

In [70]:
def top_n_movies_user(userId, n):
   user_similarities = pd.DataFrame(cosine_similarity(user_pivot), columns=user_pivot.index, index=user_pivot.index)  
   weights = (user_similarities.query("userId!=@user_id")[user_id] / sum(user_similarities.query("userId!=@user_id")[user_id]))  
   not_rated_movies = user_pivot.loc[user_pivot.index!=user_id, user_pivot.loc[user_id,:]==0]
   weighted_averages = pd.DataFrame(not_rated_movies.T.dot(weights), columns=["predicted_rating"])
   recommendations = weighted_averages.merge(movies, left_index=True, right_on="movieId")
   top_movies=recommendations.sort_values("predicted_rating", ascending=False).head()
   return top_movies

In [71]:
top_n_movies_user(2,10)

predicted_rating  movieId  \
277           2.654727      318   
507           2.087327      589   
659           1.859548      858   
2078          1.663564     2762   
3638          1.624820     4993   

                                                  title                genres  
277                    Shawshank Redemption, The (1994)           Crime|Drama  
507                   Terminator 2: Judgment Day (1991)         Action|Sci-Fi  
659                               Godfather, The (1972)           Crime|Drama  
2078                            Sixth Sense, The (1999)  Drama|Horror|Mystery  
3638  Lord of the Rings: The Fellowship of the Ring,...     Adventure|Fantasy

In [76]:
#using user_pivot
def top_n_movies_user(userId, n):
    movie_ratings = user_pivot[userId]
    similar_to_movies = user_pivot.corrwith(movie_ratings)
    corr_movies = pd.DataFrame(similar_to_movies, columns=['PearsonR'])
    corr_movies.dropna(inplace=True)
    movies_corr_summary = corr_movies.join(rating['rating_count'])
    movies_corr_summary.drop(userId, inplace=True) # drop the inputed restaurant itself
    top_10= movies_corr_summary[movies_corr_summary['rating_count']>=10].sort_values('PearsonR', ascending=False).head(5)
    top_10= top_10.merge(movies, left_index=True, right_on="movieId")
    result=list(top_10['title'][0:5])
    return result


In [77]:
top_n_movies_user(2,5)

['Lion King, The (1994)',
 'Mrs. Doubtfire (1993)',
 'Mask, The (1994)',
 'Nightmare Before Christmas, The (1993)',
 'Home Alone (1990)']

In [80]:
#using movie_rating_pivot
def top_n_movies_user(userId, n):
    movie_ratings = movie_rating_pivot[userId]
    similar_to_movies = movie_rating_pivot.corrwith(movie_ratings)
    corr_movies = pd.DataFrame(similar_to_movies, columns=['PearsonR'])
    corr_movies.dropna(inplace=True)
    movies_corr_summary = corr_movies.join(rating['rating_count'])
    movies_corr_summary.drop(userId, inplace=True) # drop the inputed restaurant itself
    top_10= movies_corr_summary[movies_corr_summary['rating_count']>=10].sort_values('PearsonR', ascending=False).head(5)
    top_10= top_10.merge(movies, left_index=True, right_on="movieId")
    result=list(top_10['title'][0:5])
    return result

In [79]:
top_n_movies_user(2,5)

/usr/local/lib/python3.7/dist-packages/numpy/lib/function_base.py:2683: RuntimeWarning: Degrees of freedom <= 0 for slice
  c = cov(x, y, rowvar, dtype=dtype)
/usr/local/lib/python3.7/dist-packages/numpy/lib/function_base.py:2542: RuntimeWarning: divide by zero encountered in true_divide
  c *= np.true_divide(1, fact)


['Ice Storm, The (1997)',
 'Hate (Haine, La) (1995)',
 'Seven Years in Tibet (1997)',
 'One Fine Day (1996)',
 'Three Colors: Red (Trois couleurs: Rouge) (1994)']

check jupyter lab for streamlit app output
file names: item_based, user_based, Streamlitapp

# The final overall code is in item_based.py
